### PR TITLE
Added JSON serialization support for enums

### PR DESCRIPTION
**Pull request by bilgili**

### CODE REVIEW

File: zerobuf/Schema.h [hunk 1]
```diff
@@ -46,11 +46,13 @@ struct Schema
...
                                allocator for dynamic arrays */
-        FIELD_ISSTATIC    //!< bool if the field is static or not
+        FIELD_ISSTATIC,   //!< bool if the field is static or not
+        FIELD_ISENUM,     //!< bool if type is an enumeration
+        FIELD_ISNESTED    //!< bool if type is a nested class
     };
 
     /** Contains the necessary information to describe one object field. */
-    typedef std::tuple< std::string, std::string, size_t, size_t, bool > Field;
+    typedef std::tuple< std::string, std::string, size_t, size_t, bool, bool, bool > Field;
```
Comment (eile) [inline zerobuf/Schema.h]: missing one bool?

**Comment by eile** [comment]
Looks good to me, although @tribal-tec should have a look after the merge

**Comment by bilgili** [comment]
Do not merge it (yet), I think I found a nice way to generalize JSONizing

**Comment by bilgili** [comment]
@hernando : I am fixing all at once. There would not be need for a warning.

File: zerobuf/Zerobuf.cpp [hunk 2]
```diff
@@ -41,19 +41,29 @@ void Zerobuf::setZerobufData( const void* data, size_t size )
 std::string Zerobuf::toJSON() const
 {
     Json::Value rootJSON;
-    for( const auto& valueSchema : getSchema().fields )
-        addValueToJSON( rootJSON, valueSchema )
+    for( auto valueSchema : getSchema().fields )
```
Comment (tribal-tec) [inline zerobuf/Zerobuf.cpp]: why removed?

File: zerobuf/NonMovingSubAllocator.h [hunk 3]
```diff
@@ -17,7 +17,9 @@ class NonMovingSubAllocator : public NonMovingBaseAllocator
 {
 public:
     ZEROBUF_API NonMovingSubAllocator( NonMovingBaseAllocator* parent,
-                                       size_t index, size_t numDynamic,
+                                       size_t offset,
```
Comment (tribal-tec) [inline zerobuf/NonMovingSubAllocator.h]: why reformatted?

File: zerobuf/Zerobuf.h [hunk 4]
```diff
@@ -10,6 +10,7 @@
 #include <zerobuf/api.h>
 #include <zerobuf/Types.h>
 #include <servus/uint128_t.h>
+#include <zerobuf/jsoncpp/json/json.h>
```
Comment (tribal-tec) [inline zerobuf/Zerobuf.h]: was an internal dependency so far, why public needed now?

File: zerobuf/detail/JsonSerialization.h [hunk 5]
```diff
@@ -36,6 +37,14 @@ Json::Value _toJSON< servus::uint128_t,
     return json;
 }
 
+/** Specialization for servus::uint128_t, represented as an object in JSON */
```
Comment (tribal-tec) [inline zerobuf/detail/JsonSerialization.h]: uint128_t not handled here

File: tests/testSchema.fbs [hunk 6]
```diff
@@ -1,6 +1,16 @@
 
 namespace test;
 
+enum TestEnum : uint {
```
Comment (tribal-tec) [inline tests/testSchema.fbs]: the 'base class' is needed or uint should be default w/o need to specify it?

Comment (bilgili) [inline tests/testSchema.fbs]: It applies the standard schema from FlatBuffers, we also discussed but no conclusion, we can assume in our parser it is uint.

File: zerobuf/Zerobuf.h (hunk 4 above)
Comment (bilgili) [inline zerobuf/Zerobuf.h]: Thought of returning Json::values from ZeroBuf objects already knows how to read as json themselves. And, construct the Json tree at one from the main object asked to be read/write as json. My ideas are not concrete on anything yet. If you have an idea I am ok with that too.

File: zerobuf/detail/JsonSerialization.h (hunk 5 above)
Comment (bilgili) [inline zerobuf/detail/JsonSerialization.h]: Yep, thats wrong.

File: zerobuf/Zerobuf.cpp (hunk 2 above)
Comment (bilgili) [inline zerobuf/Zerobuf.cpp]: it should be const.

File: zerobuf/Zerobuf.h (hunk 4 above)
Comment (tribal-tec) [inline zerobuf/Zerobuf.h]: toJSON returns JSON as string. Converting this to Json::Value outside of ZeroBuf should be possible.

File: zerobuf/Zerobuf.cpp (hunk 2 above)
Comment (tribal-tec) [inline zerobuf/Zerobuf.cpp]: Still a copy then. If there is no real reason to change that, keep const ref please.

File: zerobuf/Zerobuf.h (hunk 4 above)
Comment (bilgili) [inline zerobuf/Zerobuf.h]: json to string to json does not seem proper to me. All to JSON tree than a single convert to string seems more covenient.

File: zerobuf/Zerobuf.cpp (hunk 2 above)
Comment (bilgili) [inline zerobuf/Zerobuf.cpp]: I know that.

File: zerobuf/Zerobuf.h (hunk 4 above)
Comment (tribal-tec) [inline zerobuf/Zerobuf.h]: I would keep it internally then, as it is also used internally only now. The implementation of the string version should reuse the Json::Value one (it's C&P now). For symmetry, add fromJSON( Json::Value )?

Comment (bilgili) [inline zerobuf/Zerobuf.h]: I was playing with things. It will move to the private implementation later.

**Comment by eile** [comment]
@bilgili : I changed the unit test to reflect how nested classes should behave imo (taking a copy of a const nested class should create a copy of the data) - I'll review the rest, maybe you can start fixing this?

See here: https://github.com/eile/ZeroBuf/commit/319a77329c5404b4ad8ce9d9f6962b294e596c0a#diff-a898963f90daa1d8885593282b1c14f2L81

**Comment by bilgili** [comment]
@eile : I thought about that but later I decided not to, because we want cheap zero copies.

File: zerobuf/Zerobuf.cpp [hunk 7]
```diff
@@ -6,61 +6,327 @@
...
+
+        const Schema::SchemaFunction& schemaFunc =
+                std::get< Schema::FIELD_SCHEMAFUNC >( valueSchema );
+
+        const Schema& schema = schemaFunc();
+
+        const NonMovingBaseAllocator* parentAllocator =
+                   static_cast< const NonMovingBaseAllocator* >( allocator );
+
+        ConstVector< Zerobuf > values( parentAllocator, offset, schema.staticSize );
```
Comment (eile) [inline zerobuf/Zerobuf.cpp]: That only works by pure luck. Zerobuf is pure virtual, and can't be allocated. You really only need the number of elements, can't you get that somewhere else?

Comment (bilgili) [inline zerobuf/Zerobuf.cpp]: I did it.

**Comment by hernando** [comment]
I'm struggling to understand the original implementation and the changes. This PR became so huge that even the title is wrong, this is implementing more than just enums.
Isn't it possible to break it down into smaller ones?
I have two suggestions. I first PR could do the PImpl idiom for Zerobuf, a second one could add the enum support. Other PRs could deal with the allocators as stand alone classes and provide unit tests for them.
I don't know if this workflow is easy with Github, but with Gerrit I find it really useful to have a collection of sequential commits and then do fixups on specific commits with rebase as I fix design decisions later on.

**Comment by bilgili** [comment]
I understand you pain but Iit is a bit difficult to separate things now. The idea was to implement  the necessary things for doing the transfer function. And, I had to change things and things all around that. I believe even after this PR we will need to change things. We can discuss if you want to understand more.

File: zerobuf/BaseVector.h [hunk 8]
```diff
@@ -29,15 +31,53 @@ class BaseVector
...
+        const size_t dynOff = (size_t)_parent->template getDynamicPtr<uint8_t>( _index )
+                              - (size_t)_parent->getData();
+
+        ConstNonMovingSubAllocator* constRef =
+                new ConstNonMovingSubAllocator( static_cast<const NonMovingBaseAllocator*>( _parent ),
+                                                dynOff + index * _elemSize,
+                                                0,
+                                                _elemSize );
+
+        const Q constRefObj( constRef );
```
Comment (bilgili) [inline zerobuf/BaseVector.h]: I did not like the code below ( needs implicit knowledge of there is a copyable buffer ). Lets discuss what will be the nicer way of having it. A way to do that is creating a new NonMovableBuffer and copying the prev buffer through a copy constructor.

**Comment by eile** [comment]
I'm reviewing this and trying to simplify this as much as possible. The allocation code got out of hand, but I'm not sure yet how to reign it back in. Nobodys fault really, just this PR pushed it over the edge. Since I started digging this hole I feel responsible for it...

File: zerobuf/BaseVector.h (hunk 8 above)
Comment (bilgili) [inline zerobuf/BaseVector.h]: I did that way :)

**Comment by bilgili** [comment]
We can discuss if you want. I did not want to change things so much to keep it simple on my way. What we can do is keep this commit as is. And, keep the functionality same by modifying and simplifying things.

**Comment by eile** [comment]
Merged in #17

**Pull request closed without merging**